### PR TITLE
Kafka v2

### PR DESCRIPTION
**Pull request by jochman**
## Status
Ready

## Description
Added new V2 Kafka integration. programmed in python.

## Screenshots
Paste here any images that will help the reviewer

## Does it break backward compatibility?
   - No 
Old integration had no EntryContest at all

## Must have
- [x] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Additional changes
Describe additional changes done, for example adding a function to common server.

### CODE REVIEW

File: Integrations/integration-Kafka_V2.yml [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+    - name: topic
+      required: true
+      description: A Topic to filter by
+    - name: value
+      required: true
+      description: Message value in string format
+    - name: partitioning-key
+      description: Message Partition key
+    description: Publish a message to Kafka
+  - name: kafka-consume-msg
```
[the remark this answers is not in the document]
Comment (jochman) [reply Integrations/integration-Kafka_V2.yml]: shouldn't we reserve the old commands? if not, I'll change all other command names. If we should use them both (and deprecate the old ones) let me know

File: Integrations/integration-Kafka_V2.yml [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+
+    def produce_message():
+        """
+        Producing message to kafka topic
+        """
+        topic = demisto.args().get('topic')
+        value = demisto.args().get('value')
+        partitioning_key = demisto.args().get('partitioning-key')
+
+        partitioning_key = str(partitioning_key)
```
[the remark this answers is not in the document]
Comment (jochman) [reply Integrations/integration-Kafka_V2.yml]: the values are getting back in unicode which have no todigit functionm, so I prefered to make them str.

File: Integrations/integration-Kafka_V2.yml [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+        if partitioning_key.isdigit():
+            partitioning_key = int(partitioning_key)
+        else:
+            partitioning_key = None
+
+        if topic in KAFKA_CLIENT.topics:
+            kafka_topic = KAFKA_CLIENT.topics[topic]
+            with kafka_topic.get_sync_producer() as producer:
+                producer.produce(
```
[the remark this answers is not in the document]
Comment (jochman) [reply Integrations/integration-Kafka_V2.yml]: yeah, on the try/except block

File: Integrations/integration-Kafka_V2.yml [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+    def consume_message():
+        """
+        Consuming one message from topic
+        """
+        topic = demisto.args().get('topic')
+        offset = demisto.args().get('offset')
+        partition = demisto.args().get('partition', None)
+
+        # Checking if valid offset is given
+        offset = str(offset)
```
[the remark this answers is not in the document]
Comment (jochman) [reply Integrations/integration-Kafka_V2.yml]: offset can get back as number, unicode or str. I'm just going for the safe coding

File: Integrations/integration-Kafka_V2.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Itay4) [reply Integrations/integration-Kafka_V2.yml]: it does have `isdigit()`

File: Integrations/integration-Kafka_V2.yml [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+                consumer = kafka_topic.get_simple_consumer(
+                    auto_offset_reset=last_fetch,
+                    reset_offset_on_start=True
+                )
+                for i in range(last_fetch, latest_offset):
+                    message = consumer.consume()
+                    incidents.append(create_incident(message=message, topic=topic))
+                    if message.offset == last_fetch - 1:
+                        break;
+                demisto.setLastRun({'last_offset': latest_offset})
```
[the remark this answers is not in the document]
Comment (jochman) [reply Integrations/integration-Kafka_V2.yml]: exactly

File: Integrations/integration-Kafka_V2.yml [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+        if partition.isdigit():
+            partition = int(partition)
+        else:
+            partition = None
+
+        incidents = list()
+
+        # Chck for topic in Kafka
+        if topic in KAFKA_CLIENT.topics:
+            last_offset = demisto.getLastRun().get('last_offset')
```
[the remark this answers is not in the document]
Comment (jochman) [reply Integrations/integration-Kafka_V2.yml]: `last_fetch = last_offset if last_offset > offset else offset` this take care of it.

File: Integrations/integration-Kafka_V2.yml [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+  type: python
+  commands:
+  - name: kafka-print-topics
+    arguments: []
+    outputs:
+    - contextPath: Kafka.Topic.Name
+      description: Kafka topic name
+      type: string
+    description: Print all partitions for a topic
+  - name: kafka-publish-msg
```
[the remark this answers is not in the document]
Comment (jochman) [reply Integrations/integration-Kafka_V2.yml]: here for bc

**Pull request merged**